core/jvm/src/BufferExtJvm.kt

```diff
@@ -0,0 +1,204 @@
+    val maxToCopy = minOf(remainingByteCount, Segment.SIZE - tail.limit).toInt()
+    val bytesRead = input.read(tail.data, tail.limit, maxToCopy)
+    if (bytesRead == -1) {
+      if (tail.pos == tail.limit) {
+        // We allocated a tail segment, but didn't end up needing it. Recycle!
+        head = tail.pop()
+        SegmentPool.recycle(tail)
+      }
+      if (forever) return
+      throw EOFException("Stream exhausted before $byteCount bytes were read.")
+    }
+    tail.limit += bytesRead
+    size += bytesRead.toLong()
+    remainingByteCount -= bytesRead.toLong()
+  }
+}
+
+/**
+ * Writes [byteCount] bytes from this buffer to [out].
+ *
+ * @param out the [OutputStream] to write to.
+ * @param byteCount the number of bytes to be written, [Buffer.size] by default.
+ *
+ * @throws IllegalArgumentException when [byteCount] is negative or exceeds the buffer size.
+ */
+public fun Buffer.writeTo(out: OutputStream, byteCount: Long = size) {
+  checkOffsetAndCount(size, 0, byteCount)
+  var remainingByteCount = byteCount
+
+  var s = head
+  while (remainingByteCount > 0L) {
+    val toCopy = minOf(remainingByteCount, s!!.limit - s.pos).toInt()
+    out.write(s.data, s.pos, toCopy)
+
+    s.pos += toCopy
+    size -= toCopy.toLong()
+    remainingByteCount -= toCopy.toLong()
+
+    if (s.pos == s.limit) {
+      val toRecycle = s
+      s = toRecycle.pop()
+      head = s
+      SegmentPool.recycle(toRecycle)
+    }
+  }
+}
+
+/**
+ * Copy bytes from this buffer's subrange, starting at [startIndex] and ending at [endIndex], to [out].
+ *
+ * @param out the destination to copy data into.
+ * @param startIndex the index (inclusive) of the first byte to copy, `0` by default.
+ * @param endIndex the index (exclusive) of the last byte to copy, `buffer.size` by default.
+ *
+ * @throws IndexOutOfBoundsException when [startIndex] or [endIndex] is out of range of buffer indices.
+ * @throws IllegalArgumentException when `startIndex > endIndex`.
+ */
+public fun Buffer.copyTo(
+  out: OutputStream,
+  startIndex: Long = 0L,
+  endIndex: Long = size
+) {
+  checkBounds(size, startIndex, endIndex)
+  if (startIndex == endIndex) return
+
+  var currentOffset = startIndex
+  var remainingByteCount = endIndex - startIndex
+
+  // Skip segments that we aren't copying from.
+  var s = head
+  while (currentOffset >= s!!.limit - s.pos) {
+    currentOffset -= (s.limit - s.pos).toLong()
+    s = s.next
+  }
+
+  // Copy from one segment at a time.
+  while (remainingByteCount > 0L) {
+    val pos = (s!!.pos + currentOffset).toInt()
+    val toCopy = minOf(s.limit - pos, remainingByteCount).toInt()
+    out.write(s.data, pos, toCopy)
+    remainingByteCount -= toCopy.toLong()
+    currentOffset = 0L
+    s = s.next
+  }
+}
+
+/**
+ * Writes up to [ByteBuffer.remaining] bytes from this buffer to the sink.
+ * Return the number of bytes written.
+ *
+ * @param sink the sink to write data to.
+ */
+public fun Buffer.readAtMostTo(sink: ByteBuffer): Int {
```

Comment:
Im a bit confused with interop functions here, both naming and what should be provided.
we have `readAtMostTo(sink, byteCount)` and `readAtMost(array, start, end)` + we have `readTo(array)` (which will read to size) and others
May be it make sense to recheck on which functions (and their names) we should provide for interop (in separate issue?)
May be it's ok now, though, for now it's a harder to understand them comparing to `common` API

Reply:
https://github.com/Kotlin/kotlinx-io/issues/143#issuecomment-1604232178